File: src/Dignite.FlexFields.Site/Dignite/FlexFields/Site/Matrix/MatrixBlockType.cs
using System.Collections.Generic;

namespace Dignite.FlexFields.Site.Matrix;

/// <summary>
/// One block type a <c>Matrix</c> field's configuration declares - a named, admin-authored schema
/// (<see cref="Fields"/>) that a block instance (<see cref="MatrixBlockValue"/>) can be an occurrence of.
/// Mirrors DynamicForms' own <c>MatrixBlockType</c>.
/// </summary>
public class MatrixBlockType
{
    /// <summary>
    /// Stable key a <see cref="MatrixBlockValue.BlockTypeName"/> refers back to. Not the display label -
    /// renaming a block type without updating already-stored blocks orphans them the same way renaming a
    /// top-level field does (see <c>FlexFieldDictionaryExtensions.RenameField</c>'s own remarks); nothing
    /// in this field type migrates that for you.
    /// </summary>
    public string Name { get; set; } = default!;

    public string DisplayName { get; set; } = default!;

    public List<InlineFieldDefinition> Fields { get; set; } = new();
}

File: angular/projects/site/src/lib/field-types/matrix/matrix-block-type.ts
import type { InlineFieldDefinition } from '../inline-field-definition';
import { normalizeInlineFieldDefinitions } from '../inline-field-definition';

/** One block type a `Matrix` field's configuration declares. Mirrors `MatrixBlockType`. */
export interface MatrixBlockType {
  name: string;
  displayName: string;
  fields: InlineFieldDefinition[];
}

/** One block instance - what a `Matrix` field's value is a list of. Mirrors `MatrixBlockValue`. */
export interface MatrixBlockValue {
  blockTypeName: string;
  values: Record<string, unknown>;
}

/** Reads a stored `Matrix.BlockTypes` configuration value, defensively. */
export function normalizeMatrixBlockTypes(source: unknown): MatrixBlockType[] {
  if (!Array.isArray(source)) {
    return [];
  }

  return source.map(item => ({
    name: (item?.name as string) ?? '',
    displayName: (item?.displayName as string) ?? '',
    fields: normalizeInlineFieldDefinitions(item?.fields),
  }));
}

/** Reads a stored Matrix field's value - a list of block instances - defensively. */
export function normalizeMatrixBlockValues(source: unknown): MatrixBlockValue[] {
  if (!Array.isArray(source)) {
    return [];
  }

  return source.map(item => {
    const value = (item ?? {}) as Partial<MatrixBlockValue>;
    return {
      blockTypeName: value.blockTypeName ?? '',
      values: (value.values as Record<string, unknown>) ?? {},
    };
  });
}
